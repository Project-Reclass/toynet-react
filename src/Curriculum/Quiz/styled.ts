/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import { Button, Text } from '@chakra-ui/react';
import styled from '@emotion/styled';

export const QuizScore = styled(Text)`
  color: ${({ percent, minScore }:{percent: number, minScore: number}) => (
    percent >= minScore ? '#39a139' : '#ff5151'
  )};
  margin: 1.5rem auto;
`;

export const SubmitQuiz = styled(Button)`
  margin: 0 auto;
`;

export const CheckIcon = styled('p')`
  margin-right: 1rem;
  &:after {
    content: '✔';
    color: #90EE90;
  }
`;

export const IncorrectIcon = styled('p')`
  margin-right: 1rem;
  &:after {
    content: '✘';
    color: red;
  }
`;

export const QuestionLabel = styled(Text)`
  color: ${({ isIncorrect }:{isIncorrect: boolean}) => (
    isIncorrect ? '#C5A3B4' : '#FFFFFF'
  )};
`;

export const QuizContainer = styled('div')`
  margin: 0 auto;
  width: 80vw;
  padding: 50px;
`;

export const AnswerContainer = styled('div')`
  margin: 0 auto;
  width: 50vw;
`;
