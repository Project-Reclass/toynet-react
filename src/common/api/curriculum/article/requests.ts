/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import axios from 'axios';

export interface ArticleImage {
  file: string;
  credit: string;

  caption?: string;
}

export interface ArticleMeta {
  source: string;
  title: string;
  author: string;
  images: ArticleImage[];
}

export const getArticleText = async (articleId: number): Promise<string> => {
  const { data } = await axios.get(`/data/article/${articleId}/article.md`);
  return data;
};

export const getArticleMeta = async (articleId: number): Promise<ArticleMeta> => {
  const { data } = await axios.get(`/data/article/${articleId}/meta.json`);
  return data;
};