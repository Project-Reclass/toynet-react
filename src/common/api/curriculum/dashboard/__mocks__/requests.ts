/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/

import { DashboardIntf } from 'src/common/types/curriculum';

const defaultMockData = {
  id: 1,
  name: 'test',
  introduction: 'intro',
  modules: [],
};

export let mockData: DashboardIntf = {...defaultMockData};

export function __resetMock() {
  mockData = {...defaultMockData};
}

export function __setMockData(data: DashboardIntf) {
  mockData.id = data.id;
  mockData.name = data.name;
  mockData.introduction = data.introduction;
  mockData.modules = data.modules;
}

export const getCurriculum = async (): Promise<DashboardIntf | undefined> => {
  return Promise.resolve(mockData);
};