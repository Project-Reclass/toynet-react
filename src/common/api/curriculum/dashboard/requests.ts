import axios from 'axios';
import { DashboardIntf } from 'src/common/types/curriculum';

export const getCurriculum = async (curriculumId: number): Promise<DashboardIntf | undefined> => {
  const { data } = await axios.get(`/data/curriculum/${curriculumId}.json`);
  return data;
};