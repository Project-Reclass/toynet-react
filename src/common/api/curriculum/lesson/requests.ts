import axios from 'axios';

type LessonSlideURI = string;
export interface Meta {
  numSlides: number;
}

export const getLessonSlide = async (submoduleId: number, slideId: number): Promise<LessonSlideURI> => {
  const { data } = await axios.get(`/lesson/${submoduleId}/${slideId}.png`);
  return data;
};

export const getLessonMeta = async (submoduleId: number): Promise<Meta | undefined> => {
  const { data } = await axios.get(`/lesson/${submoduleId}/meta.json`);
  return data;
};
