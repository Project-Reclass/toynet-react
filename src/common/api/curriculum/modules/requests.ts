import { simulateNetwork } from 'src/common/utils';
import modules from 'src/__data__/curriculum/modules.json';

export const getAllModules = () => {
  return simulateNetwork(() => modules);
};

export const getModuleById = (id: number) => {
  return simulateNetwork(() => modules.find(module => module.id === id));
};

export const getModuleByName = (name: string) => {
  return simulateNetwork(() => modules.find(module => module.name === name));
};