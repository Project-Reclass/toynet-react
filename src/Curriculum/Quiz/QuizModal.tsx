/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import React, { useMemo } from 'react';
import {
  Modal,
  ModalOverlay,
  ModalContent,
  ModalCloseButton,
  ModalBody,
  ModalFooter,
  Button,
  Heading,
  Text,
  Flex,
} from '@chakra-ui/react';
import { CheckIcon, IncorrectIcon, QuizScore } from './styled';

const MIN_SCORE = 0.7;

interface Props {
  total: number;
  isOpen: boolean;
  numCorrect: number;

  done: () => any;
  tryAgain: () => any;
}

const QuizModal = ({ total, numCorrect, isOpen, done, tryAgain}: Props) => {
  const percentage = useMemo(() => numCorrect / total, [numCorrect, total]);

  return (
    <Modal isOpen={isOpen} onClose={done}>
      <ModalOverlay />
      <ModalContent backgroundColor='#33373b'>
        <ModalCloseButton />

        <ModalBody textAlign='center'>
          <Heading>Quiz Score</Heading>
          <Text>{percentage >= MIN_SCORE ?
            'Great Job!' :
            'So close! Let\'s try again.'}
          </Text>
          <QuizScore
            fontSize='3xl'
            percent={percentage}
            minScore={MIN_SCORE}
          >
            {`${(percentage * 100).toFixed(2)}%`}
          </QuizScore>
          <Flex justifyContent='space-evenly'>
            <Text display='flex' fontSize='2xl'>
              <CheckIcon />
              {numCorrect}
            </Text>
            <Text display='flex' fontSize='2xl'>
              <IncorrectIcon />
              {total - numCorrect}
            </Text>
          </Flex>
        </ModalBody>

        <ModalFooter display='flex' justifyContent='space-evenly'>
          {percentage >= MIN_SCORE &&
            <Button colorScheme="blue" mr={3} onClick={done}>
              I'm Done
            </Button>
          }
          <Button colorScheme="blue" mr={3} onClick={tryAgain}>
            Try Again
          </Button>
        </ModalFooter>
      </ModalContent>
    </Modal>
  );
};
export default QuizModal;