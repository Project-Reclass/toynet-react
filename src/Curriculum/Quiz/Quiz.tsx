/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import { useState } from 'react';
import { useParams } from 'react-router-dom';
import {
  SimpleGrid,
  Box,
  Flex,
  Text,
  useDisclosure,
  Divider,
  Heading,
  Stack,
  Container,
  Button,
  Radio,
  RadioGroup,
} from '@chakra-ui/react';

import { StringMap } from 'src/common/types';
import { useQuiz } from 'src/common/api/curriculum/quiz';
import { Question } from 'src/common/api/curriculum/quiz/requests';
import LoadingContainer from 'src/common/components/LoadingContainer';
import NavigationWithDivider from 'src/common/components/NavigationWithDivider';

import {
  CheckIcon,
  IncorrectIcon,
  QuestionLabel,
  QuizContainer,
  AnswerContainer,
} from './styled';
import QuizModal from './QuizModal';

interface Params {
  moduleId: string;
  quizId: string;
}

const answeredIndices = new Set();

const Quiz = () => {
  const { moduleId, quizId } = useParams<Params>();
  const { data, isLoading } = useQuiz(Number(moduleId), Number(quizId));
  const { isOpen, onClose, onOpen } = useDisclosure({ defaultIsOpen: false });

  const [isQuizSubmitted, setIsQuizSubmitted] = useState<boolean>(false);
  const [answerIsCorrect, setAnswerIsCorrect] = useState<StringMap>({});


  const handleAnsweredQuestion = (q: Question, qIndex: number, answerPicked: number) => {
    return () => {
      answeredIndices.add(qIndex);
      answerIsCorrect[qIndex] = q.answer === answerPicked;
      setAnswerIsCorrect(answerIsCorrect);
    };
  };

  const submitQuiz = () => {
    if (data?.items.length !== answeredIndices.size) {
      alert('You answered ' + answeredIndices.size + ' out of ' + data?.items.length + ' questions!');
    } else {
      onOpen();
      setIsQuizSubmitted(true);
    };
  };

  // We refresh the page so that the radio boxes are cleared.
  // This is because there is not simple way to clear radio
  // boxes programmatically.
  const resetQuiz = () => {
    // Add a `#` to the href to ensure that the screen scrolls to the top on refresh
    if (!window.location.href.endsWith('#'))
      window.location.href = `${window.location.href}#`;

    window.location.reload();
  };

  const numCorrect = (
    Object.values(answerIsCorrect || {})
      .filter(Boolean).length
  );

  return (
    <QuizContainer id="#">
      <LoadingContainer isLoading={isLoading}>
        <Container maxW='container.xl'>
          <SimpleGrid columns={1} spacing={5}>
            <Stack spacing={2}>
              <Heading size="lg">{`Survey: ${data?.name}`}</Heading>
              <Text size='sm' color='whiteAlpha.800'>{data?.description}</Text>
              <Divider />
            </Stack>
            {data?.items && data.items.map((q: Question, qIndex: number) => (
              <Box px={5} py={2} color="white" key={q.question}>
                <Flex>
                  {isQuizSubmitted && (answerIsCorrect[qIndex] ?
                    <CheckIcon color="green.500"/> : <IncorrectIcon color="red.500" />)}
                  <Text fontSize='lg'>{ `${qIndex + 1}. ${q.question}`}</Text>
                </Flex>
                <AnswerContainer>
                  <RadioGroup>
                    <SimpleGrid columns={2} spacingX={1}>
                      {q.options.map((option, optionIndex) => (
                        <Radio
                          key={option}
                          id={option}
                          color='white'
                          data-testid={option}
                          name={qIndex.toString()}
                          value={option}
                          onChange={handleAnsweredQuestion(q, qIndex, optionIndex)}
                          margin='5px'
                        >
                          <QuestionLabel
                            as={'label'}
                            key={option}
                            fontSize='lg'
                            isIncorrect={isQuizSubmitted &&
                              !answerIsCorrect[qIndex] &&
                              q.answer === optionIndex
                            }
                          >
                            {option}
                          </QuestionLabel>
                        </Radio>
                      ))}
                    </SimpleGrid>
                  </RadioGroup>
                </AnswerContainer>
              </Box>
            ))}
          </SimpleGrid>
          <Flex my={5}>
            <Button
              colorScheme='teal'
              onClick={!isQuizSubmitted ? submitQuiz : resetQuiz}
            >
              {!isQuizSubmitted ? 'Submit Quiz' : 'Try Again'}
            </Button>
          </Flex>
          <NavigationWithDivider
            moduleId={Number(moduleId)}
            submoduleId={Number(quizId)}
            submoduleType='QUIZ'
          />
          <QuizModal
            done={onClose}
            isOpen={isOpen}
            tryAgain={resetQuiz}
            numCorrect={numCorrect}
            total={data?.items.length || 0}
          />
        </Container>
      </LoadingContainer>
    </QuizContainer>
  );
};

export default Quiz;