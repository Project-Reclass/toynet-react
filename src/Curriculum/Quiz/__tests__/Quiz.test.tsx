/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { fireEvent, waitFor, screen } from '@testing-library/react';

import Quiz from 'src/Curriculum/Quiz';
import { getQuizMeta } from 'src/common/api/curriculum/quiz/requests';
import { renderWithTheme } from 'src/common/test-utils/renderWithTheme';

jest.mock('src/common/api/curriculum/quiz/requests.ts');
const getQuizMetaMock = getQuizMeta as jest.MockedFunction<typeof getQuizMeta>;

const RenderWithRouter = ({ children, moduleId, quizId }: {children: React.ReactChild, moduleId: number, quizId: number}) => (
  <MemoryRouter initialEntries={[`/module/${moduleId}/quiz/${quizId}`]}>
    <Route path="/module/:moduleId/quiz/:quizId">{children}</Route>
  </MemoryRouter>
);

const ANSWER: number = 3;
const data = [
  {
    'question': 'How many nodes share a single channel on a bus topology?',
    'options': [
        'One node',
        'Only ones communicating',
        'Three nodes',
        'All nodes',
    ],
    'answer': ANSWER,
  },
  {
    'question': 'Which of the following is the most fault-tolerant WAN topology?',
    'options': [
        'Partial Mesh',
        'Ring',
        'Star',
        'Full-mesh',
    ],
    'answer': ANSWER,
  },
];
describe('The Quiz page', () => {

  it('should render the same based on URL parameters', () => {
    const { container } = renderWithTheme(
      <RenderWithRouter moduleId={42} quizId={64}>
        <Quiz />
      </RenderWithRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should fetch quiz data', async () => {
    getQuizMetaMock.mockResolvedValue({ items: data });

    const { getByText, getAllByText } = renderWithTheme(
      <RenderWithRouter moduleId={42} quizId={64}>
        <Quiz />
      </RenderWithRouter>,
    );

    await waitFor(() => getAllByText(/How many nodes/i));

    expect(getByText(/How many nodes/i)).toBeInTheDocument();
    expect(getByText(/fault-tolerant WAN topology/i)).toBeInTheDocument();
    expect(getByText(/Partial Mesh/i)).toBeInTheDocument();
  });

  it('should err if none answered', () => {
    jest.spyOn(window, 'alert').mockImplementation(() => {});
    getQuizMetaMock.mockResolvedValue({ items: data });
    const { getByText } = renderWithTheme(
      <RenderWithRouter moduleId={42} quizId={64}>
        <Quiz />
      </RenderWithRouter>,
    );
    const submitBtn = getByText(/submit/i);
    fireEvent.click(submitBtn);
    expect(window.alert).toBeCalled();

  });
  it('should err if not all answered', () => {
    jest.spyOn(window, 'alert').mockImplementation(() => {});
    getQuizMetaMock.mockResolvedValue({ items: data });
    const { getByText } = renderWithTheme(
      <RenderWithRouter moduleId={42} quizId={64}>
        <Quiz />
      </RenderWithRouter>,
    );
    const choice1 = screen.getByTestId('Only ones communicating');
    const submitBtn = getByText(/submit/i);
    fireEvent.click(choice1);
    fireEvent.click(submitBtn);
    expect(window.alert).toBeCalled();

  });
  it('should not err if all answered', () => {
    jest.spyOn(window, 'alert').mockImplementation(() => {});
    getQuizMetaMock.mockResolvedValue({ items: data });
    const { getByText } = renderWithTheme(
      <RenderWithRouter moduleId={42} quizId={64}>
        <Quiz />
      </RenderWithRouter>,
    );
    const choice1 = screen.getByTestId('Only ones communicating');
    const choice2 = screen.getByTestId('Partial Mesh');
    const submitBtn = getByText(/submit/i);
    fireEvent.click(choice1);
    fireEvent.click(choice2);
    fireEvent.click(submitBtn);
    expect(window.alert).not.toBeCalled();

  });

});