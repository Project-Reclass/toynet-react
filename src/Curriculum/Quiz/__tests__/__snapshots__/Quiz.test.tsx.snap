// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Quiz page should render the same based on URL parameters 1`] = `
<div>
  <div
    class="css-1cxsl7j"
    id="#"
  >
    <div
      class="chakra-container css-jf7n8r"
    >
      <div
        class="css-1yjzl2e"
      >
        <div
          class="chakra-stack css-btweb"
        >
          <h2
            class="chakra-heading css-1jb3vzl"
          >
            Survey: Quiz
          </h2>
          <p
            class="chakra-text css-1tccf60"
          >
            Mock description
          </p>
          <hr
            aria-orientation="horizontal"
            class="chakra-divider css-svjswr"
          />
        </div>
        <div
          class="css-d5qpui"
        >
          <div
            class="css-k008qs"
          >
            <p
              class="chakra-text css-i3jkqk"
            >
              1. How many nodes share a single channel on a bus topology?
            </p>
          </div>
          <div
            class="css-1ij3qf3"
          >
            <div
              class="chakra-radio-group css-0"
              role="radiogroup"
            >
              <div
                class="css-rhcq65"
              >
                <label
                  class="chakra-radio css-1081i7r"
                  data-testid="One node"
                  id="One node"
                  name="0"
                  style="margin: 5px;"
                  value="One node"
                >
                  <input
                    class="chakra-radio__input"
                    id="One node"
                    name="0"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="radio"
                    value="One node"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-radio__control css-1rggdzb"
                    data-testid="One node"
                    style="margin: 5px;"
                  />
                  <span
                    class="chakra-radio__label css-1i66d7g"
                  >
                    <label
                      class="chakra-text css-c6hj5f"
                    >
                      One node
                    </label>
                  </span>
                </label>
                <label
                  class="chakra-radio css-1081i7r"
                  data-testid="Only ones communicating"
                  id="Only ones communicating"
                  name="0"
                  style="margin: 5px;"
                  value="Only ones communicating"
                >
                  <input
                    class="chakra-radio__input"
                    id="Only ones communicating"
                    name="0"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="radio"
                    value="Only ones communicating"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-radio__control css-1rggdzb"
                    data-testid="Only ones communicating"
                    style="margin: 5px;"
                  />
                  <span
                    class="chakra-radio__label css-1i66d7g"
                  >
                    <label
                      class="chakra-text css-c6hj5f"
                    >
                      Only ones communicating
                    </label>
                  </span>
                </label>
                <label
                  class="chakra-radio css-1081i7r"
                  data-testid="Three nodes"
                  id="Three nodes"
                  name="0"
                  style="margin: 5px;"
                  value="Three nodes"
                >
                  <input
                    class="chakra-radio__input"
                    id="Three nodes"
                    name="0"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="radio"
                    value="Three nodes"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-radio__control css-1rggdzb"
                    data-testid="Three nodes"
                    style="margin: 5px;"
                  />
                  <span
                    class="chakra-radio__label css-1i66d7g"
                  >
                    <label
                      class="chakra-text css-c6hj5f"
                    >
                      Three nodes
                    </label>
                  </span>
                </label>
                <label
                  class="chakra-radio css-1081i7r"
                  data-testid="All nodes"
                  id="All nodes"
                  name="0"
                  style="margin: 5px;"
                  value="All nodes"
                >
                  <input
                    class="chakra-radio__input"
                    id="All nodes"
                    name="0"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="radio"
                    value="All nodes"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-radio__control css-1rggdzb"
                    data-testid="All nodes"
                    style="margin: 5px;"
                  />
                  <span
                    class="chakra-radio__label css-1i66d7g"
                  >
                    <label
                      class="chakra-text css-c6hj5f"
                    >
                      All nodes
                    </label>
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-d5qpui"
        >
          <div
            class="css-k008qs"
          >
            <p
              class="chakra-text css-i3jkqk"
            >
              2. Which of the following is the most fault-tolerant WAN topology?
            </p>
          </div>
          <div
            class="css-1ij3qf3"
          >
            <div
              class="chakra-radio-group css-0"
              role="radiogroup"
            >
              <div
                class="css-rhcq65"
              >
                <label
                  class="chakra-radio css-1081i7r"
                  data-testid="Partial Mesh"
                  id="Partial Mesh"
                  name="1"
                  style="margin: 5px;"
                  value="Partial Mesh"
                >
                  <input
                    class="chakra-radio__input"
                    id="Partial Mesh"
                    name="1"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="radio"
                    value="Partial Mesh"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-radio__control css-1rggdzb"
                    data-testid="Partial Mesh"
                    style="margin: 5px;"
                  />
                  <span
                    class="chakra-radio__label css-1i66d7g"
                  >
                    <label
                      class="chakra-text css-c6hj5f"
                    >
                      Partial Mesh
                    </label>
                  </span>
                </label>
                <label
                  class="chakra-radio css-1081i7r"
                  data-testid="Ring"
                  id="Ring"
                  name="1"
                  style="margin: 5px;"
                  value="Ring"
                >
                  <input
                    class="chakra-radio__input"
                    id="Ring"
                    name="1"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="radio"
                    value="Ring"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-radio__control css-1rggdzb"
                    data-testid="Ring"
                    style="margin: 5px;"
                  />
                  <span
                    class="chakra-radio__label css-1i66d7g"
                  >
                    <label
                      class="chakra-text css-c6hj5f"
                    >
                      Ring
                    </label>
                  </span>
                </label>
                <label
                  class="chakra-radio css-1081i7r"
                  data-testid="Star"
                  id="Star"
                  name="1"
                  style="margin: 5px;"
                  value="Star"
                >
                  <input
                    class="chakra-radio__input"
                    id="Star"
                    name="1"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="radio"
                    value="Star"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-radio__control css-1rggdzb"
                    data-testid="Star"
                    style="margin: 5px;"
                  />
                  <span
                    class="chakra-radio__label css-1i66d7g"
                  >
                    <label
                      class="chakra-text css-c6hj5f"
                    >
                      Star
                    </label>
                  </span>
                </label>
                <label
                  class="chakra-radio css-1081i7r"
                  data-testid="Full-mesh"
                  id="Full-mesh"
                  name="1"
                  style="margin: 5px;"
                  value="Full-mesh"
                >
                  <input
                    class="chakra-radio__input"
                    id="Full-mesh"
                    name="1"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="radio"
                    value="Full-mesh"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-radio__control css-1rggdzb"
                    data-testid="Full-mesh"
                    style="margin: 5px;"
                  />
                  <span
                    class="chakra-radio__label css-1i66d7g"
                  >
                    <label
                      class="chakra-text css-c6hj5f"
                    >
                      Full-mesh
                    </label>
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-zl0uqj"
      >
        <button
          class="chakra-button css-f6n2t5"
          type="button"
        >
          Submit Quiz
        </button>
      </div>
      <div
        class="chakra-stack css-1w6iwr8"
      >
        <hr
          aria-orientation="horizontal"
          class="chakra-divider css-svjswr"
        />
        <div
          class="css-1lads1q"
        >
          <div
            class="chakra-stack css-jqjin"
          >
            <button
              aria-label="previous"
              class="chakra-button css-1muaiii"
              disabled=""
              type="button"
            >
              <span
                class="chakra-button__icon css-1wh2kri"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-icon css-onkibi"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <g
                    fill="currentColor"
                  >
                    <path
                      d="M10.416,12a2.643,2.643,0,0,1,.775-1.875L20.732.584a1.768,1.768,0,0,1,2.5,2.5l-8.739,8.739a.25.25,0,0,0,0,.354l8.739,8.739a1.768,1.768,0,0,1-2.5,2.5l-9.541-9.541A2.643,2.643,0,0,1,10.416,12Z"
                    />
                    <path
                      d="M.25,12a2.643,2.643,0,0,1,.775-1.875L10.566.584a1.768,1.768,0,0,1,2.5,2.5L4.327,11.823a.25.25,0,0,0,0,.354l8.739,8.739a1.768,1.768,0,0,1-2.5,2.5L1.025,13.875A2.643,2.643,0,0,1,.25,12Z"
                    />
                  </g>
                </svg>
              </span>
              Previous
            </button>
            <button
              aria-label="next"
              class="chakra-button css-1muaiii"
              disabled=""
              type="button"
            >
              Next
              <span
                class="chakra-button__icon css-1hzyiq5"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-icon css-onkibi"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <g
                    fill="currentColor"
                  >
                    <path
                      d="M13.584,12a2.643,2.643,0,0,1-.775,1.875L3.268,23.416a1.768,1.768,0,0,1-2.5-2.5l8.739-8.739a.25.25,0,0,0,0-.354L.768,3.084a1.768,1.768,0,0,1,2.5-2.5l9.541,9.541A2.643,2.643,0,0,1,13.584,12Z"
                    />
                    <path
                      d="M23.75,12a2.643,2.643,0,0,1-.775,1.875l-9.541,9.541a1.768,1.768,0,0,1-2.5-2.5l8.739-8.739a.25.25,0,0,0,0-.354L10.934,3.084a1.768,1.768,0,0,1,2.5-2.5l9.541,9.541A2.643,2.643,0,0,1,23.75,12Z"
                    />
                  </g>
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
